### ISSUE TITLE
[Feature] Fire an event on events or failures

### ISSUE THREAD

**Issue by renssies** [greghesp/ha-bambulab#44]
**Describe the feature**
I'm using this home assistant plugin to inform myself and my so when the print is finished, the app is supposed to do this, but due to https://github.com/bambulab/Bambu-Handy/issues/1 this isn't working for every device... 

However the sensors only indicate state, printing or finished, not if an error has occurred or if it was paused because of spaghetti. If these events are sent over MQTT it would be great to fire an event within Home Assistant for these events. 

**Device Type:**
 - Printer: X1C
 - Firmware Version [e.g. chrome, safari]

**Comment by renssies** [comment]
The spaghetti alert I had this morning did, but a print failure (which I've only had once, and no idea why), does not. 

I believe the spaghetti alert was HMS_0C00_0300_0003_0008

For what I mean with print failure literally the notification "Print failed", this can occur when you cancel a print, but somehow also out of nowhere, in my case it was during the leveling, but it did not trigger an error in the HMS.

**Comment by renssies** [comment]
If you slice something with the PEI plate, but use the cool plate it will throw a print failure/error that is not found in the HMS.

**Comment by CitizenRacer** [comment]
Firing events on print finished, and errors would be super cool.

**Comment by AdrianGarside** [comment]
I'm not sure how to fire events in a generic way with home assistant. The notifications I get on my phone I had to write custom yaml to push them to specific devices. Is there even a generic broadcast mechanism in Home Assistant?

**Comment by npike** [comment]
Perhaps https://developers.home-assistant.io/docs/integration_events/ ?

**Comment by renssies** [comment]
I'm trying to see if I can catch the errors in MQTT, however the MQTT sends so many messages my MQTT client loses them. 

Any recommendations on a MQTT client that can work reliably for the purpose of finding out if errors are reported over MQTT?

**Comment by AdrianGarside** [comment]
If you enable debug logging for the integration and have the current dev release installed, it'll write out every received message to the log. Just be careful if running home assistant on a raspberry pi as that level of logging will kill the sd card sooner rather than later.
https://github.com/greghesp/ha-bambulab/releases/tag/v1.1.1-dev

**Comment by AdrianGarside** [comment]
That build has the new HMS sensor with error count so you can detect it going from 0 -> 1+ and ought to be able to generate your own notifications now using the data in the attributes. You can test by taking the print head cover off.

**Comment by renssies** [comment]
This seems to be working, but I would like to either see the errors move to Home assistant events. Or make the HMS Errors sensor a binary_sensor. So on = error detected, off = no error. Makes it much more readable.

**Comment by AdrianGarside** [comment]
![image](https://user-images.githubusercontent.com/26147576/233764504-f4927d70-283c-4325-8300-b1dcf64d8c06.png)

**Comment by renssies** [comment]
That's much better, this one is woring great now!

**Comment by AdrianGarside** [comment]
Now firing an even when the HMS errors clear or change:
https://github.com/greghesp/ha-bambulab/releases/tag/v1.1.1-dev13

I haven't learnt yet how this can be used in HA by the end user. It seems like I might need to add a device trigger to make this actually useful?
https://developers.home-assistant.io/docs/device_automation_trigger/

**Comment by npike** [comment]
Nice! Looks like the device trigger is the next step.

**Comment by AdrianGarside** [comment]
If I’m understanding it right, device automation and device triggers are a way to simplify creating automations by providing pre canned options. Still not sure what that looks like in practice or how to use it to create an error notification.

**Comment by AdrianGarside** [comment]
This works to send me a notification on my phone. Doesn't need/use events at all.
```
alias: P1P Encountered Errors
description: ""
trigger:
  - type: problem
    platform: device
    device_id: 3533c965f293b9a15213fb552e2aaaaf
    entity_id: binary_sensor.p1p_SERIAL_hms_errors
    domain: binary_sensor
condition: []
action:
  - service: notify.mobile_app_adrianphone
    data:
      title: P1P Encountered Errors
      message: >-
        {{ state_attr('binary_sensor.p1p_SERIAL_hms_errors', 'Count')
        }} errors encountered.
mode: single
```

**Comment by AdrianGarside** [comment]
This is triggering off an event but I haven't worked out the syntax to trigger it only for a specific device. But ideally, I think you'd want a single trigger for all your bambu printers and include which hit the error in in the notification title/message.
```
alias: P1P Encountered Errors 2
description: ""
trigger:
  - platform: event
    event_type: bambulab_hms_errors_event
condition: []
action:
  - service: notify.mobile_app_adrianphone
    data:
      title: P1P Encountered Errors 2
      message: >-
        "{{ state_attr('binary_sensor.p1p_SERIAL_hms_errors', 'Count')
        }} errors encountered: '{{
        state_attr('binary_sensor.p1p_SERIAL_hms_errors', '1-Error')
        }}'"
mode: single
```

**Comment by sophof** [comment]
I have tried to use the events to fire off warnings on my speaker, but they don't seem to work at all? The binary sensor isn't reliable, because a lot of 'errors' aren't actually errors at all and I'd like to ignore, but it is very hard to parse this using the binary sensor.

So anyway, events would be the perfect solution (assuming they also contain the relevant error in their data) but they don't appear to fire at all? I'm currently on version 1.4.

**Comment by AdrianGarside** [comment]
Yeah, Bambu chose to report lots of non-errors as errors. Such as:
'SlotX filament has run out. Please wait while old filament is purged'
and
'SlotX filament has run out and automatically switched to the slot with the same filament'

These show as unnecessarily persistent errors in the Handy app too - eventually they disappear but they've still been around well into the _next_ print for me. I may have to categorize the errors somehow to allow us to ignore the normal use ones like the above.

I hope to get some more spare time/energy soon (work got hectic) and see if I can make it easier to trigger automation from pre-canned events that you can easily select from in the home assistant UX. That part of HA development is new to me so it might go quickly or it might take me a while to work out how to do it given the HA developer docs are typically abysmal.

**Comment by sophof** [comment]
Yeah, those are persistent for me too, sometimes sticking for days (another one is "First layer inspection not possible"...). This is also why I can't parse the binary sensor well, because often the relevant error is the second or even the third and it is never reliable in which place. I'm sure I can manage it if I really try (by somehow converting it to a dict for instance), but I'd rather not.

And I only understand too well how hard it can be to find time to develop something for home assistant, even if it seems simple (and often is once you know how). What you describe would be great, but I'd say it is a nice to have for my problem, the event is the basis to build that on anyway. I've already identified the most common 'fake' errors for me and try to ignore those in my automation. The only issue now is that events are not always correctly identified because they overlap and the only the attributes are updated. 

So an event that is fired containing the error code is more than enough for me. But to reiterate, I'm not receiving any events? As far as I understand from this topic they should already be in this version right?

I do have some experience with building integrations btw, so I might be able to create a pull request this summer, but I expect it'll be dangerous to do too much processing in the integration on the error codes (they might change over time and break stuff).

**Comment by AdrianGarside** [comment]
Starting to investigate this one again to implement events that can be easily automated on top of via the UI.

**Comment by AdrianGarside** [comment]
https://github.com/greghesp/ha-bambulab/releases/tag/v1.7.0-dev2

Has print failed and an hms error event. The latter has event data payload you can use in the automation.
